res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,741 @@
+//                            from a previous position on long press)
+//
+//        * CUE/LOOP CALL - memory & delete (complex and not useful. Hot cues are sufficient)
+//
+//      * Secondary pad modes (trial attempts complex and too experimental)
+//        * Keyboard mode
+//        * Pad FX1
+//        * Pad FX2
+//        * Keyshift mode
+
+var PioneerDDJ400 = {};
+
+PioneerDDJ400.lights = {
+    beatFx: {
+        status: 0x94,
+        data1: 0x47,
+    },
+    deck1: {
+        vuMeter: {
+            status: 0xB0,
+            data1: 0x02,
+        },
+        playPause: {
+            status: 0x90,
+            data1: 0x0B,
+        },
+        shiftPlayPause: {
+            status: 0x90,
+            data1: 0x47,
+        },
+        cue: {
+            status: 0x90,
+            data1: 0x0C,
+        },
+        shiftCue: {
+            status: 0x90,
+            data1: 0x48,
+        },
+    },
+    deck2: {
+        vuMeter: {
+            status: 0xB0,
+            data1: 0x02,
+        },
+        playPause: {
+            status: 0x91,
+            data1: 0x0B,
+        },
+        shiftPlayPause: {
+            status: 0x91,
+            data1: 0x47,
+        },
+        cue: {
+            status: 0x91,
+            data1: 0x0C,
+        },
+        shiftCue: {
+            status: 0x91,
+            data1: 0x48,
+        },
+    },
+};
+
+// Store timer IDs
+PioneerDDJ400.timers = {};
+
+// Stores padmode state and encapsulates LED logic
+PioneerDDJ400.performancePads = {
+    modes: {
+        HOTCUE: 0x1B,
+        BEATLOOP: 0x6D,
+        BEATJUMP: 0x20,
+        SAMPLER: 0x22,
+        KEYBOARD: 0x69,
+        PADFX1: 0x1E,
+        PADFX2: 0x6B,
+        KEYSHIFT: 0x6F
+    },
+
+    // Start in hotcue mode
+    state: [
+        0x1B, // Channel 1
+        0x1B  // Channel 2
+    ],
+
+    // Toggles the shift specific lights for the current mode
+    toggleShiftLights: function(channel, value) {
+        if (this.state[channel] === this.modes.BEATJUMP) {
+            midi.sendShortMsg(channel === 0 ? 0x98 : 0x9A, 0x26, value);
+            midi.sendShortMsg(channel === 0 ? 0x98 : 0x9A, 0x27, value);
+        }
+    },
+
+    // Clear all pad mode lights and reset back to hotcue state
+    reset: function() {
+        for (var m in this.modes) {
+            if (this.modes[m] !== this.modes.HOTCUE) {
+                midi.sendShortMsg(0x90, this.modes[m], 0x00);
+                midi.sendShortMsg(0x91, this.modes[m], 0x00);
+            }
+        }
+        this.state[0] = this.state[1] = this.modes.HOTCUE;
+
+        midi.sendShortMsg(0x90, this.modes.HOTCUE, 0x7f);
+        midi.sendShortMsg(0x91, this.modes.HOTCUE, 0x7f);
+    }
+};
+
+// Save the Shift State
+PioneerDDJ400.shiftState = [0, 0];
+
+// JogWheel
+PioneerDDJ400.vinylMode = true;
+PioneerDDJ400.alpha = 1.0/8;
+PioneerDDJ400.beta = PioneerDDJ400.alpha/32;
+PioneerDDJ400.highspeedScale = 150; // multiplier for fast seek through track using SHIFT+JOGWHEEL
```

Comment:
move comment to line above

Reply:
I think `fastSeekScale` would be a more self-explanatory name.

Reply:
agreed. changed as suggested